src/index.js

```diff
@@ -3,52 +3,55 @@ import React from 'react'
 const changedArray = (a = [], b = []) =>
   a.length !== b.length || a.some((item, index) => !Object.is(item, b[index]))
 
-const initialState = {error: null, info: null}
+const initialState = {error: null}
 class ErrorBoundary extends React.Component {
   static getDerivedStateFromError(error) {
     return {error}
   }
 
   state = initialState
+  updatedWithError = false
   resetErrorBoundary = (...args) => {
     this.props.onReset?.(...args)
+    this.reset()
+  }
+
+  reset() {
+    this.updatedWithError = false
     this.setState(initialState)
   }
 
   componentDidCatch(error, info) {
-    this.props.onError?.(error, info?.componentStack)
-    this.setState({info})
+    this.props.onError?.(error, info)
   }
 
   componentDidUpdate(prevProps) {
-    const {error, info} = this.state
+    const {error} = this.state
     const {resetKeys} = this.props
-    if (
-      error !== null &&
-      info !== null &&
-      changedArray(prevProps.resetKeys, resetKeys)
-    ) {
+
+    // There's an edge case where if the thing that triggered the error
+    // happens to *also* be in the resetKeys array, we'd end up resetting
+    // the error boundary immediately. This would likely trigger a second
+    // error to be thrown.
+    // So we make sure that we don't check the resetKeys on the first call
+    // of cDU after the error is set
+    if (error !== null && !this.updatedWithError) {
```

Comment:
I don't think I fully understand why the `resetKeys` feature is helpful, or what the purpose of `onResetKeysChange` does– but the way you're using the `updatedWithError` instance attribute looks safe (since it's only modified during the commit phase or during an imperative reset "action".

Reply:
Yeah, it's actually quite helpful. Here's an example of it in action: https://react-hooks.netlify.app/6 (It's the Extra Credit 8, which you can play with here: https://react-hooks.netlify.app/isolated/final/06.extra-8.js)

Basically, if there's a property that you anticipate resulting in an error to be thrown, then you can put it in the `resetKeys` array and after an error has been thrown, if any values in the array changes then it will be reset automatically.

Here's a gif that demonstrates what this does. With the reset keys, the error boundary is automatically reset when I change the pokemon name:

![Screen Recording 2020-09-11 at 9 15 39 AM](https://user-images.githubusercontent.com/1500684/92943094-933a9c00-f40f-11ea-8424-9626816db28b.gif)

Reply:
Thanks for elaborating. The concept seems a bit odd to me, but not necessarily in a *bad* way. It could just be because it's a different way of thinking about using the boundary than would have occurred to me.

I guess it's roughly similar to using an id-based `key` or something to completely reset a subtree between view changes.

Anyway :smile: it doesn't look like it would cause problems. (Instance properties can sources of subtle side effects but I think this one looks fine.)